### PR TITLE
Add PI-DeepONet example for 1D Poisson equation.

### PR DESCRIPTION
**Pull request by samuelburbulla**
This PR proposes an example for training a physics-informed DeepONet with DeepXDE.
The example solves a simple one-dimensional Poisson equation where the source term is the input to the solution operator.

Related to #1308.

### CODE REVIEW

**Comment by samuelburbulla** [comment]
I improved this PR a little more:

- Incorporate some fixes for using PyTorch backend
- Add auxiliary_variables in PyTorch.NN
- Use L-BFGS for training the DeepONet

Here's a plot of the result:

pideeponet_poisson1d

File: deepxde/data/pde_operator.py [hunk 1]
```diff
@@ -275,10 +276,12 @@ def losses_test(self, targets, outputs, loss_fn, inputs, model, aux=None):
     def train_next_batch(self, batch_size=None):
         if self.train_x is None:
             func_feats = self.func_space.random(self.num_func)
-            func_vals = self.func_space.eval_batch(func_feats, self.eval_pts)
+            func_vals = self.func_space.eval_batch(
+                func_feats, self.eval_pts
+            ).astype(config.real(np))
```
Comment (lululxvi) [inline deepxde/data/pde_operator.py:281]: Don't change this. `self.func_space.eval_batch` should return the correct `dtype`; if not, then should modify `self.func_space.eval_batch` or the inputs `func_feats` and `self.eval_pts`.

File: deepxde/data/pde_operator.py [hunk 2]
```diff
@@ -275,10 +276,12 @@ def losses_test(self, targets, outputs, loss_fn, inputs, model, aux=None):
...
             vx = self.func_space.eval_batch(
                 func_feats, self.pde.train_x[:, self.func_vars]
-            )
+            ).astype(config.real(np))
```
Comment (lululxvi) [inline deepxde/data/pde_operator.py:284]: The same as above.

File: deepxde/data/pde_operator.py [hunk 3]
```diff
@@ -262,7 +262,8 @@ def _losses(self, outputs, loss_fn, inputs, model, num_func):
             losses.append(losses_i)
 
         losses = zip(*losses)
-        losses = [bkd.reduce_mean(bkd.as_tensor(l)) for l in losses]
+        # Reduce mean without loosing derivatives
+        losses = [sum(l) / len(l) for l in losses]
```
Comment (lululxvi) [inline deepxde/data/pde_operator.py:266]: What is the difference between `reduce_mean` and `sum / len`?

**Comment by lululxvi** [comment]
I suggest splitting this PR into two separate PRs
1. the demo example code;
2. Pytorch backed code.

**Comment by samuelburbulla** [comment]
Thanks, @lululxvi, for your review!

This PR is now split into:
- #1311 (this one) including the example only
- #1318 including the PyTorch backend modifications

I have found better solutions for the points mentioned, see comments above.

File: deepxde/data/pde_operator.py (hunk 1 above)
Comment (samuelburbulla) [reply deepxde/data/pde_operator.py]: Okay, I added `@to_tensor` decorators to the function space `eval_batch` functions.

File: deepxde/data/pde_operator.py (hunk 3 above)
Comment (samuelburbulla) [reply deepxde/data/pde_operator.py]: The issue is that `bkd.tensor(l)` looses gradients of `l` for PyTorch, due to the use of `torch.to_tensor` for a tuple of tensors.

I proposed an adaptation of `torch.tensor()` that uses `torch.stack` for list/tuple of tensors. Then, this line can stay as it is.

**Comment by sibyjackgrove** [comment]
@samuelburbulla Thanks for making this very useful example. @lululxvi would you be merging this soon?

File: deepxde/data/function_spaces.py [hunk 4]
```diff
@@ -90,6 +91,7 @@ def random(self, size):
     def eval_one(self, feature, x):
         return np.dot(feature, x ** np.arange(self.N))
 
+    @return_tensor
```
Comment (lululxvi) [inline deepxde/data/function_spaces.py:94]: Don't return tensor. This file is designed to be pure Python/Numpy.

**Comment by lululxvi** [comment]
Which backends does the new code support? The existing code supports tensorflow 1.x.

**Comment by samuelburbulla** [comment]
> Which backends does the new code support? The existing code supports tensorflow 1.x.

This example is tested with tensorflow.compat.v1, tensorflow, and pytorch.

File: deepxde/model.py [hunk 5]
```diff
@@ -281,7 +281,8 @@ def outputs(training, inputs):
             return self.net(inputs)
 
         def outputs_losses(training, inputs, targets, auxiliary_vars, losses_fn):
-            self.net.auxiliary_vars = auxiliary_vars
+            if auxiliary_vars is not None:
+                self.net.auxiliary_vars = torch.as_tensor(auxiliary_vars).requires_grad_()
```
Comment (lululxvi) [inline deepxde/model.py:285]: Why you need the grad of auxiliary_vars?

**Comment by lululxvi** [comment]
I have added a few PI-DeepONet examples, see https://deepxde.readthedocs.io/en/latest/demos/operator.html#pi-deeponet

Could you split this PR into two PRs:
1. only modify the source code;
2. add new example

**Comment by samuelburbulla** [comment]
> I have added a few PI-DeepONet examples

Nice!

> Could you split this PR into two Pls

Done, see #1379.

**Comment by samuelburbulla** [comment]
Rebased on master.

Together with #1403, this example works with all mentioned backends.

Is there anything else that needs to be done here?

File: examples/operator/pideeponet_poisson1d.py [hunk 6]
```diff
@@ -0,0 +1,102 @@
+"""Backend supported: tensorflow.compat.v1, tensorflow, pytorch"""
+import numpy as np
+import deepxde as dde
+import matplotlib.pyplot as plt
```
Comment (lululxvi) [inline examples/operator/pideeponet_poisson1d.py:4]:
```
import deepxde as dde
import matplotlib.pyplot as plt
import numpy as np
```

File: examples/operator/pideeponet_poisson1d.py [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+    [num_eval_points, 32, p],
+    [dim_x, 32, p],
+    activation="tanh",
+    kernel_initializer="Glorot normal",
+)
+
+# Define and train model
+model = dde.Model(pde_op, net)
+
+if dde.backend.backend_name == "pytorch":
```
Comment (lululxvi) [inline examples/operator/pideeponet_poisson1d.py:66]: Why different optimizers for different backends?

**Comment by lululxvi** [comment]
Reformat via black https://pypi.org/project/black/

File: docs/demos/operator/poisson_1d_pideeponet.rst [hunk 8]
```diff
@@ -0,0 +1,114 @@
+Physics-informed DeepONet for Poisson equation in 1D
```
Comment (lululxvi) [inline docs/demos/operator/poisson_1d_pideeponet.rst:1]: Just "Poisson equation in 1D"
